File: app/src/main/java/com/dbeginc/dbweather/ui/main/weather/WeatherTabFragment.java
package com.dbeginc.dbweather.ui.main.weather;

import android.app.AlertDialog;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.SearchManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.graphics.drawable.VectorDrawableCompat;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.NotificationCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;

import com.dbeginc.dbweather.R;
import com.dbeginc.dbweather.databinding.WeatherTabLayoutBinding;
import com.dbeginc.dbweather.models.datatypes.geonames.GeoName;
import com.dbeginc.dbweather.models.datatypes.weather.Alert;
import com.dbeginc.dbweather.models.datatypes.weather.HourlyData;
import com.dbeginc.dbweather.models.datatypes.weather.WeatherData;
import com.dbeginc.dbweather.ui.BaseFragment;
import com.dbeginc.dbweather.ui.animation.ForegroundToBackgroundTransformer;
import com.dbeginc.dbweather.ui.main.weather.adapters.CustomFragmentAdapter;
import com.dbeginc.dbweather.ui.main.weather.adapters.HourAdapter;
import com.dbeginc.dbweather.utils.helper.ColorManager;
import com.github.clans.fab.FloatingActionButton;
import com.google.android.gms.ads.AdRequest;

import java.util.List;
import java.util.Locale;

import io.reactivex.subjects.PublishSubject;

import static com.dbeginc.dbweather.utils.holder.ConstantHolder.LOCATION_UPDATE;
import static com.dbeginc.dbweather.utils.holder.ConstantHolder.WEATHER_ALERT_ID;
import static com.dbeginc.dbweather.utils.holder.ConstantHolder.WEATHER_INFO_KEY;

/**
 * Created by darel on 28.05.17.
 * Weather Activity
 */

public class WeatherTabFragment extends BaseFragment implements IWeatherView, SearchView.OnSuggestionListener, ViewPager.OnPageChangeListener {
    private static final int LAST_PAGE_INDEX = 6;
    private WeatherTabLayoutBinding layoutBinding;
    private WeatherPresenter presenter;
    private CustomFragmentAdapter fragmentAdapter;
    private HourAdapter hourAdapter;
    private WeatherData weatherData;
    private GeoName selectedLocation;
    private Drawable floatingButtonIcon;
    private String gpsLocationName;
    private ComponentName componentName;
    private int floatingButtonCount = 1;
    private final Handler handler = new Handler();
    private final ColorManager colorManager = ColorManager.getInstance();
    private static final String COMPONENT_NAME_KEY = "COMPONENT_NAME_KEY";
    private final ViewGroup.LayoutParams floatingButtonLayoutParams = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT);

    public static synchronized WeatherTabFragment newInstance(@NonNull final WeatherData weatherData) {
        final WeatherTabFragment weatherTabFragment = new WeatherTabFragment();
        final Bundle args = new Bundle();
        args.putParcelable(WEATHER_INFO_KEY, weatherData);
        weatherTabFragment.setArguments(args);
        return weatherTabFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = new WeatherPresenter(this, mAppDataProvider);

        final Bundle arguments = getArguments();
        if (savedInstanceState != null) {
            final WeatherData data = savedInstanceState.getParcelable(WEATHER_INFO_KEY);
            weatherData = data != null ? data : weatherData;
            componentName = savedInstanceState.getParcelable(COMPONENT_NAME_KEY);

        } else if (arguments != null) {
            final WeatherData data = arguments.getParcelable(WEATHER_INFO_KEY);
            weatherData = data != null ? data : weatherData;
            if (presenter.isCurrentWeatherFromGps() || gpsLocationName == null) {
                gpsLocationName = weatherData.getWeatherInfoList().get(0).locationName.get();
            }
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        componentName = getActivity().getComponentName();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(WEATHER_INFO_KEY, weatherData);
        outState.putParcelable(COMPONENT_NAME_KEY, componentName);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        layoutBinding = DataBindingUtil.inflate(inflater, R.layout.weather_tab_layout, container, false);
        return layoutBinding.getRoot();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        handler.post(() -> setupLocationLookupFeature(view.getContext()));
        handler.post(this::setupAds);
        setupViewPager(weatherData);
        setupHourlyRecyclerView(weatherData.getHourlyWeatherList());


        final int backgroundColor = colorManager.getBackgroundColor(weatherData.getWeatherInfoList().get(0).icon.get());
        layoutBinding.nsvContainer.setBackgroundResource(backgroundColor);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            layoutBinding.nsvContainer.setNestedScrollingEnabled(true);
            layoutBinding.nsvContainer.setVerticalScrollBarEnabled(true);
            layoutBinding.nsvContainer.setScrollbarFadingEnabled(true);
            layoutBinding.nsvContainer.setSmoothScrollingEnabled(true);
        }
        layoutBinding.viewPager.setPageTransformer(false, new ForegroundToBackgroundTransformer());
        layoutBinding.civLeftButton.setOnClickListener(v -> goToPreviousPage());
        layoutBinding.civRightButton.setOnClickListener(v -> goToNextPage());
        layoutBinding.civLeftButton.setRippleColor(resources.getColor(R.color.colorAccent));
        layoutBinding.civRightButton.setRippleColor(resources.getColor(R.color.colorAccent));
        layoutBinding.srlContainer.setOnRefreshListener(() -> handler.post(() -> {
            if (presenter.isCurrentWeatherFromGps()) { presenter.getWeather(); }
            else {
                presenter.getWeatherForCity(selectedLocation);
                presenter.setCurrentWeatherFromGps(false);
            }
        }));
        handleBackButtonVisibility(0);

        setupFloatingMenu();
        handler.post(presenter::loadUserCities);
        presenter.subscribeToEvents();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (!presenter.isFirstRun() && isNetworkAvailable() && presenter.isCurrentWeatherFromGps()) {
            presenter.getWeather();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        presenter.clearState();
    }

    @Override
    public void setupLocationLookupFeature(@NonNull final Context context) {
        final SearchManager searchManager = (SearchManager) context.getSystemService(Context.SEARCH_SERVICE);
        layoutBinding.searchLocationView.setSearchableInfo(searchManager.getSearchableInfo(componentName));
        layoutBinding.searchLocationView.setIconifiedByDefault(false);
        layoutBinding.searchLocationView.setSubmitButtonEnabled(false);
        layoutBinding.searchLocationView.setOnSuggestionListener(this);
    }

    @Override
    public void showWeather(@NonNull final WeatherData weatherData) {
        this.weatherData = weatherData;
        fragmentAdapter.updateFragments(weatherData.getWeatherInfoList());
        if (hourAdapter != null) { hourAdapter.updateData(weatherData.getHourlyWeatherList()); }
        if (weatherData.getAlertList() != null) { showWeatherAlerts(weatherData.getAlertList()); }
        final int backgroundColor = colorManager.getBackgroundColor(weatherData.getWeatherInfoList().get(0).icon.get());
        layoutBinding.nsvContainer.setBackgroundResource(backgroundColor);
        layoutBinding.srlContainer.setRefreshing(false);
        weatherDataUpdateEvent.onNext(weatherData);
        if (presenter.isCurrentWeatherFromGps()) {
            gpsLocationName = this.weatherData.getWeatherInfoList().get(0).locationName.get();
            layoutBinding.currentLocationMenuItem.setLabelText(gpsLocationName);
        }
    }

    @Override
    public void showWeatherAlerts(@NonNull List<Alert> listWeatherAlert) {
        final NotificationManager notificationManager = (NotificationManager) getActivity().getSystemService(Context.NOTIFICATION_SERVICE);

        for (final Alert alert : listWeatherAlert) {
            final NotificationCompat.Builder builder = new NotificationCompat.Builder(getActivity());

            final Intent browserIntent = new Intent()
                    .setAction(Intent.ACTION_VIEW)
                    .setData(Uri.parse(alert.getUri().isEmpty() ? "https://github.com/404" : alert.getUri()))
                    .addCategory(Intent.CATEGORY_BROWSABLE);

            final int notificationId = WEATHER_ALERT_ID;

            builder.setSmallIcon(R.mipmap.ic_launcher);
            builder.setContentTitle(alert.getTitle());
            builder.setContentText(alert.getDescription());
            builder.setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION));
            builder.setCategory(NotificationCompat.CATEGORY_ALARM);
            builder.setContentIntent(PendingIntent.getActivity(getActivity(), notificationId, browserIntent, PendingIntent.FLAG_UPDATE_CURRENT));
            builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE);

            builder.setContentInfo(String.format(Locale.getDefault(),
                    getString(R.string.notification_alert_city)
                    , weatherData.getWeatherInfoList().get(0).locationName.get()));

            notificationManager.notify(notificationId, builder.build());
        }
    }

    @Override
    public Context getAppContext() { return super.getAppContext(); }

    @Override
    public PublishSubject<String> getLocationUpdateEvent() { return locationUpdateEvent; }

    @Override
    public PublishSubject<String> getVoiceSearchEvent() { return voiceQuery; }

    @Override
    public void onLocationUpdate(@NonNull String action) {
        if (action.equalsIgnoreCase(LOCATION_UPDATE) && !presenter.isCurrentWeatherFromGps()) {
            if (isNetworkAvailable()) { presenter.getWeather(); }
            else { showNetworkNotAvailableMessage(); }
        }
    }

    /**
     * This method handle voice query event
     *
     * @param query user voice query
     */
    @Override
    public void onVoiceQueryReceived(@NonNull final String query) {
        layoutBinding.searchLocationView.setQuery(query, false);
        layoutBinding.searchLocationView.requestFocusFromTouch();
    }

    @Override
    public boolean isNetworkAvailable() { return super.isNetworkAvailable(); }

    @Override
    public void showNetworkNotAvailableMessage() {
        Snackbar.make(layoutBinding.weatherTabLayout, R.string.network_unavailable_message, Snackbar.LENGTH_LONG)
                .setActionTextColor(Color.RED)
                .show();
    }

    @Override
    public void showWeatherError() {
        Snackbar.make(layoutBinding.weatherTabLayout, getString(R.string.weather_error_message), Snackbar.LENGTH_LONG)
                .setAction(R.string.retry, v -> presenter.retryWeatherRequest())
                .setActionTextColor(Color.RED)
                .show();
    }

    @Override
    public void loadUserCities(@NonNull final List<GeoName> userCities) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            floatingButtonIcon = VectorDrawableCompat.create(resources, R.drawable.city_location_icon, getActivity() != null ? getActivity().getTheme() : null);
        } else {
            floatingButtonIcon = ResourcesCompat.getDrawable(resources, R.drawable.city_location_icon, getActivity() != null ? getActivity().getTheme() : null);
        }

        final int locationSize = userCities.size() > 6 ? 6 : userCities.size();

        for (int index = 0; index < locationSize; index++ ) {
            addFloatingButtonToMenu(userCities.get(index));
        }
    }

    private void addFloatingButtonToMenu(GeoName location) {
        final FloatingActionButton button = new FloatingActionButton(getAppContext());
        button.setLabelText(String.format(Locale.getDefault(), "%s, %s", location.getName(), location.getCountryName()));
        button.setImageDrawable(floatingButtonIcon);
        button.setColorNormal(Color.WHITE);
        button.setColorPressedResId(R.color.colorPrimary);
        button.setColorRippleResId(R.color.colorPrimaryDark);
        button.setLayoutParams(floatingButtonLayoutParams);
        button.setOnClickListener(v -> handler.post(() -> {
            presenter.setCurrentWeatherFromGps(false);
            presenter.doWeSaveWeather(true);
            presenter.getWeatherForCity(location);
            selectedLocation = location;
            layoutBinding.srlContainer.setRefreshing(true);
            layoutBinding.floatingLocationsMenu.close(true);
        }));
        layoutBinding.floatingLocationsMenu.addMenuButton(button);
        floatingButtonCount++;
    }

    @Override
    public void notifySuccessfullAddedLocation() {
        Snackbar.make(layoutBinding.weatherTabLayout, R.string.successfully_added_city, Snackbar.LENGTH_LONG)
                .setActionTextColor(Color.WHITE)
                .show();
    }

    @Override
    public void addLocationToMenu(@NonNull final GeoName location) { addFloatingButtonToMenu(location); }

    @Override
    public boolean onSuggestionSelect(int i) { return true; }

    @Override
    public boolean onSuggestionClick(int i) {
        selectedLocation = mListOfLocation.get(i);

        final DialogInterface.OnClickListener cancelLocationClick = (dialog, which) -> dialog.cancel();
        final DialogInterface.OnClickListener displayClick = (dialog, which) -> handler.post(() -> {
            presenter.getWeatherForCity(selectedLocation);
            presenter.setCurrentWeatherFromGps(false);
            presenter.doWeSaveWeather(false);
            handler.postAtTime(this::closeSoftKeyboard, 5000);
            layoutBinding.srlContainer.setRefreshing(true);
        });
        final DialogInterface.OnClickListener addToFavorite = (dialog, which) -> handler.post(() -> {
            presenter.doWeSaveWeather(true);
            presenter.setCurrentWeatherFromGps(false);
            presenter.getWeatherForCity(selectedLocation);
            presenter.addToFavorite(selectedLocation);
            if (floatingButtonCount < 7) {
                presenter.isLocationInDatabase(
                        String.format(Locale.getDefault(), "%s, %s", selectedLocation.getName(), selectedLocation.getCountryName()),
                        selectedLocation
                );
            }
            handler.postAtTime(this::closeSoftKeyboard, 5000);
            layoutBinding.srlContainer.setRefreshing(true);
        });

        new AlertDialog.Builder(getActivity())
                .setMessage(String.format(Locale.getDefault(),
                        getAppContext().getString(R.string.alert_add_location_text),
                        selectedLocation.getName()))
                .setNegativeButton(android.R.string.cancel, cancelLocationClick)
                .setNeutralButton(R.string.display, displayClick)
                .setPositiveButton(R.string.display_add, addToFavorite)
                .create()
                .show();
        return true;
    }


    @Override
    public void onPageScrolled(int i, float v, int i1) {}

    @Override
    public void onPageSelected(int i) {
        handleBackButtonVisibility(i);
        handleNextButtonVisibility(i);
    }

    @Override
    public void onPageScrollStateChanged(int i) {}

    private void setupAds() {
        final AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice(AdRequest.DEVICE_ID_EMULATOR)
                .addTestDevice("687D1ACC5C0ACF7F698DBA9A4E258FFA")
                .addTestDevice("C20BB1C5369BFDFD4992ED89CD62F271")
                .build();
        layoutBinding.adVMain.loadAd(adRequest);
    }

    private void setupFloatingMenu() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            layoutBinding.currentLocationMenuItem.setImageDrawable(VectorDrawableCompat.create(resources, R.drawable.current_location_icon, getActivity() != null ? getActivity().getTheme() : null));
            layoutBinding.floatingLocationsMenu.getMenuIconView().setImageDrawable(VectorDrawableCompat.create(resources, R.drawable.add_location_icon, getActivity() != null ? getActivity().getTheme() : null));
        } else {
            layoutBinding.currentLocationMenuItem.setImageDrawable(ResourcesCompat.getDrawable(resources,
                    R.drawable.current_location_icon, getActivity() != null ? getActivity().getTheme() : null));
            layoutBinding.floatingLocationsMenu.getMenuIconView()
                    .setImageDrawable(ResourcesCompat.getDrawable(resources, R.drawable.add_location_icon, getActivity() != null ? getActivity().getTheme() : null));
        }

        layoutBinding.currentLocationMenuItem.setOnClickListener(v -> handler.post(() -> {
            presenter.setCurrentWeatherFromGps(true);
            presenter.doWeSaveWeather(false);
            presenter.getWeather();
            layoutBinding.srlContainer.setRefreshing(true);
            layoutBinding.floatingLocationsMenu.close(true);
        }));

        layoutBinding.floatingLocationsMenu.setOnMenuToggleListener(b -> {
            if (b) { handleMenuButtonClicked(); }
            else {
                final int currentItem = layoutBinding.viewPager.getCurrentItem();
                handleBackButtonVisibility(currentItem);
                handleNextButtonVisibility(currentItem);
            }
        });
        layoutBinding.currentLocationMenuItem.setLabelText(gpsLocationName);
    }

    private void handleMenuButtonClicked() {
        if (layoutBinding.civRightButton.getVisibility() == View.VISIBLE) {
            layoutBinding.civRightButton.setVisibility(View.INVISIBLE);
        }
        if (layoutBinding.civLeftButton.getVisibility() == View.VISIBLE) {
            layoutBinding.civLeftButton.setVisibility(View.INVISIBLE);
        }
    }

    private void handleBackButtonVisibility(final int position) {
        if (position == 0) {
            layoutBinding.civLeftButton.setVisibility(View.INVISIBLE);
            layoutBinding.civLeftButton.setEnabled(false);

        } else {
            layoutBinding.civLeftButton.setVisibility(View.VISIBLE);
            layoutBinding.civLeftButton.setEnabled(true);
        }
    }

    private void handleNextButtonVisibility(final int position) {
        if (position == LAST_PAGE_INDEX) {
            layoutBinding.civRightButton.setVisibility(View.INVISIBLE);
            layoutBinding.civRightButton.setEnabled(false);

        } else {
            layoutBinding.civRightButton.setVisibility(View.VISIBLE);
            layoutBinding.civRightButton.setEnabled(true);
        }
    }

    /**
     * Handle The Back Navigation Button Clicked
     */
    private void goToPreviousPage() {
        layoutBinding.viewPager.setCurrentItem(layoutBinding.viewPager.getCurrentItem() - 1, true);
    }

    private void goToNextPage() {
        layoutBinding.viewPager.setCurrentItem(layoutBinding.viewPager.getCurrentItem() + 1);
    }

    /**
     * Setup Hourly RecyclerView
     * @param hourlyData a list of hourly weather data to display
     */
    private void setupHourlyRecyclerView(@NonNull final List<HourlyData> hourlyData) {
        hourAdapter = new HourAdapter(hourlyData, presenter.getRxSubscriptions());
        layoutBinding.hourlyRecyclerView.setHasFixedSize(true);
        layoutBinding.hourlyRecyclerView.setAdapter(hourAdapter);
        layoutBinding.hourlyRecyclerView.setLayoutManager(new LinearLayoutManager(getAppContext(),
                LinearLayoutManager.HORIZONTAL, false));
    }

    /**
     * Setup Weather ViewPager
     * @param weatherData weather Data
     */
    private void setupViewPager(WeatherData weatherData) {
        fragmentAdapter = new CustomFragmentAdapter(getChildFragmentManager(),
                weatherData.getWeatherInfoList());

        layoutBinding.weatherTabLayout.setBackgroundColor(colorManager.getBackgroundColor(weatherData.getWeatherInfoList().get(0).icon.get()));
        layoutBinding.viewPager.setAdapter(fragmentAdapter);
        layoutBinding.viewPager.addOnPageChangeListener(this);
    }

    private void closeSoftKeyboard() {
        final InputMethodManager inputMethodManager = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        layoutBinding.searchLocationView.setQuery("", false);
        layoutBinding.searchLocationView.setIconified(false);
        layoutBinding.searchLocationView.clearFocus();
        inputMethodManager.hideSoftInputFromWindow(layoutBinding.searchLocationView.getWindowToken(), 0);
    }
}

File: app/src/main/java/com/dbeginc/dbweather/weather/view/WeatherTabFragment.kt
/*
 *  Copyright (C) 2017 Darel Bitsy
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License
 */

package com.dbeginc.dbweather.weather.view

import android.app.*
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.constraint.ConstraintLayout
import android.support.design.widget.Snackbar
import android.support.graphics.drawable.VectorDrawableCompat
import android.support.v4.app.NotificationCompat
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SearchView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.dbeginc.dbweather.R
import com.dbeginc.dbweather.base.BaseFragment
import com.dbeginc.dbweather.databinding.FragmentWeatherTabBinding
import com.dbeginc.dbweather.utils.animations.widgets.RainFallView
import com.dbeginc.dbweather.utils.animations.widgets.SnowFallView
import com.dbeginc.dbweather.utils.broadcastreceivers.NotificationCanceler
import com.dbeginc.dbweather.utils.helper.ColorManager
import com.dbeginc.dbweather.utils.helper.LocationObserver
import com.dbeginc.dbweather.utils.holder.ConstantHolder
import com.dbeginc.dbweather.utils.holder.ConstantHolder.*
import com.dbeginc.dbweather.utils.utility.*
import com.dbeginc.dbweatherweather.viewmodels.AlertWeatherModel
import com.dbeginc.dbweatherweather.viewmodels.LocationWeatherModel
import com.dbeginc.dbweatherweather.viewmodels.WeatherModel
import com.dbeginc.dbweather.weather.WeatherTabContract
import com.dbeginc.dbweather.weather.adapters.daily.DayAdapter
import com.dbeginc.dbweather.weather.adapters.hourly.HourAdapter
import com.github.clans.fab.FloatingActionButton
import java.util.*
import javax.inject.Inject

/**
 * Created by darel on 22.09.17.
 *
 * Weather Tab Fragment
 */
class WeatherTabFragment : BaseFragment(), WeatherTabContract.WeatherTabView, SearchView.OnSuggestionListener {
    @Inject lateinit var presenter: WeatherTabContract.WeatherTabPresenter

    private val floatingButtonLayoutParams by lazy { ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT) }
    private val locationIcon by lazy { getLocationsIcon() }
    private val currentLocationIcon by lazy { currentLocationIcon() }
    private val locationMenuIcon by lazy { getLocationsMenuIcon() }
    private val animationLayoutParams by lazy { getLayoutParameter() }

    private lateinit var dailyWeatherAdapter: DayAdapter
    private lateinit var hourlyWeatherAdapter: HourAdapter
    private lateinit var binding: FragmentWeatherTabBinding
    private lateinit var componentName: ComponentName
    private lateinit var locationChangeEvent: LocationObserver
    private var weatherData: WeatherModel? = null

    override fun onCreate(savedState: Bundle?) {
        super.onCreate(savedState)
        Injector.injectWeatherTabDep(this)
        if (savedState != null) {
            weatherData = savedState.getParcelable(WEATHER_INFO_KEY)
            componentName = savedState.getParcelable(COMPONENT_NAME_KEY)
        }

        hourlyWeatherAdapter = HourAdapter(weatherData?.hourly ?: emptyList())

        dailyWeatherAdapter = DayAdapter(weatherData?.daily ?: emptyList())

        if (applicationPreferences.isGpsOn()) {
            locationChangeEvent = LocationObserver(activity!!)
            locationChangeEvent.observe(this, android.arch.lifecycle.Observer<android.location.Location> { location -> handleLocationUpdate(location) })
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        componentName = activity!!.componentName
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(WEATHER_INFO_KEY, weatherData)
        outState.putParcelable(COMPONENT_NAME_KEY, componentName)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(
                inflater.cloneInContext(android.view.ContextThemeWrapper(activity, R.style.AppTheme_Main_WeatherTab)),
                R.layout.fragment_weather_tab,
                container,
                false
        )
        return binding.root
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        presenter.bind(this)
    }

    override fun onDestroyView() {
        super.onDestroyView()

        cleanState()
    }

    override fun onSuggestionSelect(position: Int): Boolean {
        presenter.onUserLocationSelected(position)
        return true
    }

    override fun onSuggestionClick(position: Int): Boolean {
        presenter.onUserLocationSelected(position)
        return true
    }

    /********************************* View Part *********************************/

    override fun setupView() {
        setupLocationsMenu()

        setupDailyWeather(dailyLayoutManager =LinearLayoutManager(context,LinearLayoutManager.VERTICAL,false))

        setupHourlyWeather(hourlyLayoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false))

        binding.searchLocationView.setupLookupFeature()

        binding.srlContainer.setOnRefreshListener { askForWeather() }

        presenter.loadUserCities()

        askForWeather()
    }

    override fun cleanState() = presenter.unBind()

    override fun displayWeather(weather: WeatherModel) {
        weatherData = weather

        binding.currentWeatherLayout?.current = weatherData!!.current

        dailyWeatherAdapter.updateData(weatherData!!.daily)

        hourlyWeatherAdapter.updateData(weatherData!!.hourly)

        if (weatherData!!.alerts != null) {
            showWeatherAlerts(weatherData!!.alerts!!)
        }

        binding.weatherTabLayout.setBackgroundResource(ColorManager.getBackgroundColor(weatherData!!.current.icon))

        when(weather.current.icon) {
            R.drawable.rain -> binding.weatherTabLayout.showRainingAnimation()
            R.drawable.snow -> binding.weatherTabLayout.showSnowFallAnimation()
            else -> binding.weatherTabLayout.removeWeatherAnimation()
        }
    }

    override fun getGpsLatitude(): Double = applicationPreferences.getGpsLatitude()

    override fun getGpsLongitude(): Double = applicationPreferences.getGpsLongitude()

    override fun getGpsLocation(): String = applicationPreferences.getGpsLocation()

    override fun defineGpsCoordinates(location: LocationWeatherModel) = applicationPreferences.updateGpsCoordinates(location.name, location.latitude, location.longitude)

    override fun defineLastPreviewedCoordinates(location: LocationWeatherModel) {
        applicationPreferences.updateLastPreviewedCoordinates(
                location.name.plus(",${location.countryName},${location.countryCode}".format()),
                location.latitude,
                location.longitude
        )
    }

    override fun showUserLocations(locations: List<LocationWeatherModel>) {

        val locationToNotDisplay = (binding.floatingLocationsMenu.childCount.minus(1) downTo 0)
                .mapNotNull { binding.floatingLocationsMenu.getChildAt(it) }
                .filter { view -> view is FloatingActionButton }
                .mapNotNull { view -> (view as FloatingActionButton).labelText }

        locations.filterNot {
            location -> locationToNotDisplay.contains(location.name.plus(", ").plus(location.countryCode))
        }.filterIndexed { index, _ -> index <= 5 }
                .forEach { location ->
                    binding.floatingLocationsMenu.addMenuButton(
                            FloatingActionButton(context).apply {
                                labelText = location.name.plus(", ").plus(location.countryCode)
                                setImageDrawable(locationIcon)
                                colorNormal = Color.WHITE
                                setColorPressedResId(R.color.appColorPrimary)
                                setColorRippleResId(R.color.appColorPrimaryDark)
                                layoutParams = floatingButtonLayoutParams
                                setOnClickListener { presenter.getWeatherForCity(location) }
                            }
                    )
                }
    }

    override fun changeCurrentLocationSource(isGpsLocation: Boolean) = applicationPreferences.updateCurrentLocationSource(isGpsLocation)

    override fun displayLoadingWeatherStatus() {
        binding.srlContainer.isRefreshing = true
    }

    override fun hideLoadingWeatherStatus() {
        binding.srlContainer.isRefreshing = false
    }

    override fun displayUpdatingStatus() = binding.loadingStatus.show()

    override fun hideUpdatingStatus() = binding.loadingStatus.remove()

    override fun showWeatherError(message: String) {
        Snackbar.make(binding.weatherTabLayout, message, Snackbar.LENGTH_LONG)
                .setActionTextColor(Color.RED)
                .setAction(R.string.retry, { _ -> presenter.retryWeatherRequest() })
    }

    override fun showError(message: String) = binding.weatherTabLayout.snack(message, duration=Snackbar.LENGTH_LONG)

    /******************************************************************************************************************/
    private fun askForWeather() {
        if (applicationPreferences.isCurrentLocationFromGps()) presenter.getWeather() else presenter.getWeatherForCity(getUserLocation())
    }

    private fun handleLocationUpdate(location: android.location.Location?) {
        if (location != null) {
            presenter.onGpsLocationChanged(latitude=location.latitude, longitude=location.longitude)
            presenter.getWeather()
        }
    }

    private fun getUserLocation() : LocationWeatherModel {
        val location = applicationPreferences.getLastPreviewedLocation().split(",")
        val latitude = applicationPreferences.getLastPreviewedLatitude()
        val longitude = applicationPreferences.getLastPreviewedLongitude()

        return if (location.size == 1) LocationWeatherModel("", latitude, longitude, "", "")
        else LocationWeatherModel(name = location[0], countryCode=location[2], countryName=location[1], latitude=latitude, longitude=longitude)
    }

    /******************************************************************************************************************/

    private fun ConstraintLayout.showRainingAnimation() {
        val snowAnimation = findViewById<SnowFallView>(SnowFallView.VIEW_ID)

        if (snowAnimation != null) removeView(snowAnimation)

        if (findViewById<RainFallView>(RainFallView.VIEW_ID) == null) addView(RainFallView(context), animationLayoutParams)
    }

    private fun ConstraintLayout.showSnowFallAnimation() {
        val rainAnimation = findViewById<RainFallView>(RainFallView.VIEW_ID)

        if (rainAnimation != null) removeView(rainAnimation)

        if (findViewById<SnowFallView>(SnowFallView.VIEW_ID) == null) addView(SnowFallView(context), animationLayoutParams)
    }

    private fun ConstraintLayout.removeWeatherAnimation() {
        removeView(findViewById<SnowFallView>(SnowFallView.VIEW_ID))
        removeView(findViewById<RainFallView>(RainFallView.VIEW_ID))
    }

    private fun setupDailyWeather(dailyLayoutManager: LinearLayoutManager) {
        binding.dailyWeatherRCV.apply {
            adapter = dailyWeatherAdapter
            layoutManager = dailyLayoutManager
            setHasFixedSize(true)
            addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        }
    }

    private fun setupHourlyWeather(hourlyLayoutManager: LinearLayoutManager) {
        binding.hourlyRecyclerView.apply {
            adapter = hourlyWeatherAdapter
            layoutManager = hourlyLayoutManager
        }
    }

    private fun showWeatherAlerts(alerts: List<AlertWeatherModel>) {
        val notificationManager = activity?.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        initChannels(notificationManager)

        val randomize = Random().nextInt()
        val notificationSound = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
        val browserIntent = Intent()
                .setAction(Intent.ACTION_VIEW)
                .addCategory(Intent.CATEGORY_BROWSABLE)

        val cancel = getString(android.R.string.cancel)
        val openInBrowser = getString(R.string.open_with)

        for (alert in alerts) {
            val builder = NotificationCompat.Builder(activity!!, WEATHER_ALERT_CHANNEL_ID)

            browserIntent.data = if (alert.uri.isEmpty()) Uri.EMPTY else Uri.parse(alert.uri)

            val notificationId = WEATHER_ALERT_ID.xor(randomize)

            val locations = alert.regions.joinToString()


            builder.setContentInfo(getString(R.string.notification_alert_city).format(locations))
                    .setContentTitle(alert.title)
                    .setContentText(alert.description)
                    .addAction(R.drawable.ic_close_black, cancel, getDismissIntent(notificationId, context!!))
                    .addAction(R.drawable.ic_internet, openInBrowser, PendingIntent.getActivity(activity, notificationId, browserIntent, PendingIntent.FLAG_UPDATE_CURRENT))
                    .setSmallIcon(R.mipmap.ic_launcher)
                    .setSound(notificationSound)
                    .setStyle(NotificationCompat.BigTextStyle().bigText(alert.description))
                    .setCategory(NotificationCompat.CATEGORY_ALARM)
                    .setDefaults(NotificationCompat.DEFAULT_VIBRATE)
                    .setChannelId(WEATHER_ALERT_CHANNEL_ID)

            notificationManager.notify(notificationId, builder.build())
        }
    }

    private fun SearchView.setupLookupFeature() {
        val searchManager = context.getSystemService(Context.SEARCH_SERVICE) as SearchManager
        setSearchableInfo(searchManager.getSearchableInfo(componentName))
        setIconifiedByDefault(false)
        isSubmitButtonEnabled = false
        setOnSuggestionListener(this@WeatherTabFragment)
    }

    private fun initChannels(notificationManager: NotificationManager) {
        if (Build.VERSION.SDK_INT < 26 || notificationManager.getNotificationChannel(WEATHER_ALERT_CHANNEL_ID) != null) return

        val channel = NotificationChannel(WEATHER_ALERT_CHANNEL_ID, WEATHER_ALERT_CHANNEL_NAME, NotificationManager.IMPORTANCE_HIGH)

        channel.apply {
            description = getString(R.string.weather_alert_channel_desc)
            enableLights(true)
            lightColor = Color.RED
            enableVibration(true)
            lockscreenVisibility = Notification.VISIBILITY_PUBLIC
        }

        notificationManager.createNotificationChannel(channel)
    }

    fun onVoiceQuery(query: String?) {
        activity?.contentResolver
                ?.query(Uri.parse(ConstantHolder.SEARCH_QUERY_URI.format(query)),
                        null,
                        null,
                        null,
                        null,
                        null
                )
    }

    private fun getDismissIntent(notificationId : Int, context: Context ) : PendingIntent {
        val intent = Intent(context, NotificationCanceler::class.java)
        intent.putExtra(NOTIFICATION_KEY, notificationId)
        return PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT)
    }

    private fun setupLocationsMenu() {
        binding.floatingLocationsMenu.menuIconView.setImageDrawable(locationMenuIcon)

        binding.currentLocationMenuItem.setImageDrawable(currentLocationIcon)

        binding.currentLocationMenuItem.setOnClickListener { presenter.getWeather() }

        binding.currentLocationMenuItem.labelText = getGpsLocation()
    }

    private fun getLayoutParameter(): ConstraintLayout.LayoutParams {
        val params = ConstraintLayout.LayoutParams(ConstraintLayout.LayoutParams.WRAP_CONTENT, ConstraintLayout.LayoutParams.WRAP_CONTENT)

        params.apply {
            topToTop = binding.weatherTabLayout.id
            leftToLeft = binding.weatherTabLayout.id
            rightToRight = binding.weatherTabLayout.id
        }

        return params
    }

    private fun getLocationsMenuIcon() : Drawable {
        return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            VectorDrawableCompat.create(resources, R.drawable.add_location_icon, if (activity != null) activity!!.theme else null)!!
        } else {
            ResourcesCompat.getDrawable(resources, R.drawable.add_location_icon, if (activity != null) activity!!.theme else null)!!
        }
    }

    private fun currentLocationIcon() : Drawable {
        return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            VectorDrawableCompat.create(resources, R.drawable.current_location_icon, if (activity != null) activity!!.theme else null)!!
        } else {
            ResourcesCompat.getDrawable(resources, R.drawable.current_location_icon, if (activity != null) activity!!.theme else null)!!
        }
    }

    private fun getLocationsIcon(): Drawable {
        return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            VectorDrawableCompat.create(resources, R.drawable.city_location_icon, if (activity != null) activity!!.theme else null)!!
        } else {
            ResourcesCompat.getDrawable(resources, R.drawable.city_location_icon, if (activity != null) activity!!.theme else null)!!
        }
    }
}